sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala

```diff
@@ -28,12 +28,18 @@ class SparkOptimizer(
     experimentalMethods: ExperimentalMethods)
   extends Optimizer(catalog) {
 
-  override def batches: Seq[Batch] = (preOptimizationBatches ++ super.batches :+
+  val experimentalPreOptimizations: Seq[Batch] = Seq(Batch(
```

Comment:
also define this as `Batch` and you can use `experimentalPreOptimizations +: preOptimizationBatches` to concatenate with other batches.